pkg/prometheus/agent/daemonset.go

```diff
@@ -0,0 +1,319 @@
+// Copyright 2023 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusagent
+
+import (
+	"fmt"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/k8sutil"
+	"github.com/prometheus-operator/prometheus-operator/pkg/operator"
+	prompkg "github.com/prometheus-operator/prometheus-operator/pkg/prometheus"
+	"github.com/prometheus-operator/prometheus-operator/pkg/webconfig"
+)
+
+func makeDaemonSet(
+	name string,
+	p monitoringv1.PrometheusInterface,
+	config *prompkg.Config,
+	cg *prompkg.ConfigGenerator,
+	tlsSecrets *operator.ShardedSecret,
+) (*appsv1.DaemonSet, error) {
+	cpf := p.GetCommonPrometheusFields()
+	objMeta := p.GetObjectMeta()
+
+	if cpf.PortName == "" {
+		cpf.PortName = prompkg.DefaultPortName
+	}
+
+	// We need to re-set the common fields because cpf is only a copy of the original object.
+	// We set some defaults if some fields are not present, and we want those fields set in the original Prometheus object before building the DaemonSetSpec.
+	p.SetCommonPrometheusFields(cpf)
```

Comment:
should we factorize since we have similar in Prometheus sts and PrometheusAgent sts?

Reply:
What lines are you referring to? Line 40-49 do different things. I'm not sure we should refactor them into one function.